### DISCUSSION TITLE
Support for client_credentials grant type?

### DISCUSSION THREAD

**Discussion by cyberdynatics** [TheNetworg/oauth2-azure#166] [Ideas]
Microsoft has finally made their oAuth authentication support the client_credentials grant type, but I'm not seeing how to get it to work. I have it working in my own IMAP implementation but can't get it working using PHPMailer for SMTP.

Does anyone know if this project can be updated to work with this grant type? I'm new to PHPMailer and it's oAuth implementations so I'm not sure how to modify it myself.

**Comment by decomplexity** [comment]
MSFT's June announcement of support for  client_credentials grant type is only for IMAP and POP. Greg Taylor of the Exchange team has confirmed to me that SMTP support has not yet (Aug 5th) been released.
When it comes, the only PHPMailer changes needed are in the getToken method of the PHPMailer Oauth class, roughly on the lines of:


```
protected function getToken()
   {
   switch ($this->grantTypeValue) {
	case "authorization_code":
	// use refresh token to get access token
        return $this->provider->getAccessToken(
           $this->getGrant(),
           ['refresh_token' => $this->oauthRefreshToken]
           );
	 break;
	case "client_credentials":
        $ret = $this->provider->getAccessToken(
	       $this->grantTypeValue,  
		   ['scope' => "https://outlook.office.com/.default",
//		   ['scope' => "https://graph.microsoft.com/.default", // or maybe this, depending on how MSFT implement
		   'client_id' => $this->oauthClientId,
		   'client_secret' => $this->oauthClientSecret
		   ]);
	       return $ret;
	break;
     }
    }
```
where new property grantTypeValue will need to be passed in the instantiation of Oauth from your mainline call of PHPMailer.
We plan to add client_credentials grant type as an option to the decomplexity/SendOauth2 PHPMailer wrapper when MSFT release SMTP support.

**Reply by cyberdynatics** [reply]
Great news. Thank you for the reply and the info. I'll put this thread in my task for later.
